src/WebJobs.Script.WebHost/Management/InstanceManager.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.WebHost.Models;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Management
+{
+    public class InstanceManager : IInstanceManager
+    {
+        private static HostAssignmentContext _assignmentContext;
+        private static object _assignmentLock = new object();
+
+        private readonly ScriptHostManager _scriptHostManager;
+        private readonly WebHostSettings _webHostSettings;
+        private readonly ILogger _logger;
+
+        public InstanceManager(WebScriptHostManager scriptHostManager, WebHostSettings webHostSettings, ILoggerFactory loggerFactory)
+        {
+            _scriptHostManager = scriptHostManager;
+            _webHostSettings = webHostSettings;
+            _logger = loggerFactory.CreateLogger(nameof(InstanceManager));
+        }
+
+        public bool TryAssign(HostAssignmentContext assignmentContext)
+        {
+            if (_assignmentContext == null)
+            {
+                lock (_assignmentLock)
+                {
+                    if (_assignmentContext != null)
+                    {
+                        return _assignmentContext.Equals(assignmentContext);
+                    }
+                    _assignmentContext = assignmentContext;
+                }
+
+                // Queue the task in background, and return to caller immediately
+                Task.Run(() => Specialize(assignmentContext));
+
+                return true;
+            }
+            else
+            {
+                // No lock needed here since _assignmentContext is not null when we are here
+                return _assignmentContext.Equals(assignmentContext);
+            }
+        }
+
+        private async Task Specialize(HostAssignmentContext assignmentContext)
+        {
+            try
+            {
+                // download zip
+                var zip = assignmentContext.ZipUrl;
+
+                if (!string.IsNullOrEmpty(zip))
+                {
+                    var filePath = Path.GetTempFileName();
+
+                    await DownloadAsync(new Uri(zip), filePath);
+
+                    assignmentContext.ApplyAppSettings();
+
+                    ZipFile.ExtractToDirectory(filePath, _webHostSettings.ScriptPath, overwriteFiles: true);
+                }
+                else
+                {
+                    assignmentContext.ApplyAppSettings();
+                }
+
+                // Restart runtime.
+                _scriptHostManager.RestartHost();
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, $"Error calling {nameof(Specialize)}");
```

Comment:
What state will we be in if specialization fails? Do we want to swallow this exception?

Reply:
/cc @suwatch and @bala16 
That's a good question. I don't think that's implemented in this PR. if I remember the design correctly, we should return 503 on `/admin/instance/status`

Reply:
opened https://github.com/Azure/azure-functions-host/issues/2717 to track